usm: sharedlibraries: Handle open(2) syscall

vitkyrka opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

We currently only handle shared libraries being opened with openat(2) and openat2(2), which works fine with glibc, but musl (default on Alpine Linux, for example) uses open(2) so handle that too.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

https://datadoghq.atlassian.net/browse/USMON-989

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

Test cases are included in the PR.

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

vitkyrka committed:
usm: sharedlibraries: Handle open(2) syscall

We currently only handle shared libraries being opened with openat(2)
and openat2(2), which works fine with glibc, but musl (default on Alpine
Linux, for example) uses open(2) so handle that too.

vitkyrka committed:
usm: test: Add test for HTTPS using musl

Add a test which checks that curl HTTPS monitoring works on Alpine which
uses musl as the libc (which uses a different system call to open the
libraries as compared to glibc).

vitkyrka committed:
usm: test: Refactor HTTPS library test

Do the discovery of binary paths and shared libraries inside each test,
so that it is easier to customize, for adding a test case using Docker.

vitkyrka committed:
Merge remote-tracking branch 'origin/main' into open-syscall

guyarb commented on pkg/network/usm/monitor_tls_test.go:
```diff
@@ -94,24 +95,30 @@ func (s *tlsSuite) TestHTTPSViaLibraryIntegration() {
-				}
-			}
-		}
+		{
+			name:     "curl (musl)",
```
let's have an inline documentation

vitkyrka replied:
Added.

nplanel commented on pkg/network/usm/sharedlibraries/ebpf.go:
```diff
@@ -177,12 +179,17 @@ func sysOpenAt2Supported() bool {
 		openatProbes = append(openatProbes, openat2SysCall)
 	}
+	// arm64 doesn't have open(2)
+	if runtime.GOARCH == "amd64" {
```
may mark it explicitly like the commend
`if runtime.GOARCH != "arm64" {`

vitkyrka committed:
usm: test: Add comment with reason for musl test

vitkyrka committed:
usm: test: Prefetch correct libs

We need to run ldd inside the chroot to be able to find the correct
libs, otherwise we end up prefetching the host libs.

vitkyrka committed:
Merge remote-tracking branch 'origin/main' into open-syscall

vitkyrka committed:
usm: sharedlibraries: Adjust comment

vitkyrka commented:
/merge
